pkg/tracker/reward/reward.go

```diff
@@ -0,0 +1,412 @@
+
+	ctx, cncl := context.WithCancel(ctx)
+	return &RewardTracker{
+		client:           client,
+		logger:           logger,
+		contractInstance: contractInstance,
+		abi:              abi,
+		addr:             addr,
+		ctx:              ctx,
+		stop:             cncl,
+		tsDB:             tsDB,
+		engine:           engine,
+		aggr:             aggr,
+	}, nil
+}
+
+func (self *RewardTracker) Start() error {
+	level.Info(self.logger).Log("msg", "starting")
+
+	go self.monitorGas()
+
+	<-self.ctx.Done()
+	return nil
+}
+
+func (self *RewardTracker) Stop() {
+	self.stop()
+}
+
+func (self *RewardTracker) monitorGas() {
+	var err error
+	ticker := time.NewTicker(DefaultRetry)
+	defer ticker.Stop()
+
+	logger := log.With(self.logger, "event", "NonceSubmitted")
+
+	var sub event.Subscription
+	events := make(chan *tellor.TellorNonceSubmitted)
+	for {
+		select {
+		case <-self.ctx.Done():
+			return
+		default:
+		}
+		sub, err = self.nonceSubmittedSub(events)
+		if err != nil {
+			level.Error(logger).Log("msg", "initial subscribing to events failed")
+			<-ticker.C
+			continue
+		}
+		break
+	}
+
+	for {
+		select {
+		case <-self.ctx.Done():
+			return
+		case err := <-sub.Err():
+			if err != nil {
+				level.Error(logger).Log(
+					"msg",
+					"subscription error",
+					"err", err)
+			}
+
+			// Trying to resubscribe until it succeeds.
+			for {
+				select {
+				case <-self.ctx.Done():
+					return
+				default:
+				}
+				sub, err = self.nonceSubmittedSub(events)
+				if err != nil {
+					level.Error(logger).Log("msg", "re-subscribing to events failed", "err", err)
+					<-ticker.C
+					continue
+				}
+				break
+			}
+			level.Info(logger).Log("msg", "re-subscribed to events")
+		case event := <-events:
+			err := self.recordGasUsage(event)
+			if err != nil {
+				level.Error(self.logger).Log("msg", "record gas usage", "err", err)
+			}
+			err = self.recordGasUsageEstimated(self.ctx)
+			if err != nil {
+				level.Error(self.logger).Log("msg", "record gas usage estimation", "err", err)
+			}
+		}
+	}
+}
+
+func (self *RewardTracker) recordGasUsage(event *tellor.TellorNonceSubmitted) error {
+	receipt, err := self.client.TransactionReceipt(self.ctx, event.Raw.TxHash)
+	if err != nil {
+		return errors.Wrap(err, "receipt retrieval")
+	} else if receipt != nil && receipt.Status == types.ReceiptStatusSuccessful { // Failed transactions cost is monitored in a different process.
+		// tx, _, err := self.client.TransactionByHash(self.ctx, event.Raw.TxHash)
+		// if err != nil {
+		// 	level.Error(self.logger).Log("msg", "get transaction by hash", "err", err)
+		// 	return
+		// }
+		level.Debug(self.logger).Log("msg", "adding gas used", "gasUsed", receipt.GasUsed)
+		if err := self.addGasUsed(event.Slot.String(), receipt.GasUsed); err != nil {
+			return errors.Wrap(err, "adding gas used")
+		}
+		return nil
+	}
+	return errors.New("transaction not yet mined")
+}
+
+func (self *RewardTracker) addGasUsed(slot string, gasUsed uint64) error {
```

Comment:
lets extract this as a function in the DB package like
```
DB.Add(ctx, logger, appender, lbls , value uint64) error {
	.....
}
``

this will allow removing some duplicated code here and in few other places that add to the db.